### PR TITLE
Try to fix rare bug in reading of empty arrays

### PR DESCRIPTION
**Pull request by CurtizJ**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Try to fix rare bug while reading of empty arrays, which could lead to `Data compressed with different methods` error.

### CODE REVIEW

**Comment by CurtizJ** [comment]
I'll add a test a bit later.

**Comment by alexey-milovidov** [comment]
Very interesting...

**Comment by CurtizJ** [comment]
I couldn't construct small function test on which this bug reproduces, because it will require knowledge of too many implementation details. But it reproduces in case like this:

```sql
CREATE TABLE test_table
(
    `date` Date,
    `timestamp` UInt32,
    `host` String,
    `http_code` UInt16,
    `test_id` Array(Int32)
)
ENGINE = MergeTree(date, (timestamp, host, http_code), 8192)
```

There is one detail of data in the table: column `test_id` has huge ranges in which all arrays are empty.

```sql
SELECT *
FROM test_table
WHERE (date = '2022-01-16') AND (host = 'some_host.com') AND (test_id = [3333])
ORDER BY timestamp DESC
LIMIT 100
SETTINGS optimize_read_in_order = 1
```

In this query we read data in the reverse order of primary key. Assume that while reading we made a seek by the [default path](https://github.com/ClickHouse/ClickHouse/blob/e2364bb2dd4b8665b98402f472f6056429e4a55e/src/IO/ReadBufferFromFileDescriptor.cpp#L170-L204). We have some data in buffer from previous calls at this moment and as it written in [comment](https://github.com/ClickHouse/ClickHouse/blob/e2364bb2dd4b8665b98402f472f6056429e4a55e/src/IO/ReadBufferFromFileDescriptor.cpp#L179) we delay fetching data to the buffer to the next call of `nextImpl` . Then it happens that we are reading granule in which all arrays are empty and it will lead to call of [readBig](https://github.com/ClickHouse/ClickHouse/blob/613c9fa3c2cddbb87b33c879be6b01b5748e3fb4/src/Compression/CompressedReadBufferFromFile.cpp#L91) with `n=0`. Since `n=0` we won't call any reading  and `nextImpl` won't be triggered as we had expected. Then on next call of seek to the position, which is a bit bellow current position (remember that we are reading ranges backwards) [this check](https://github.com/ClickHouse/ClickHouse/blob/e2364bb2dd4b8665b98402f472f6056429e4a55e/src/IO/ReadBufferFromFileDescriptor.cpp#L158-L169) will pass  and we will move position in buffer without actual seek, but actually buffer contains wrong data.

File: src/IO/tests/gtest_seek_backwards.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    /// Read something to initialize the buffer.
+    in.seek(BUF_SIZE * 10, SEEK_SET);
+    readIntBinary(x, in);
+
+    /// Check 2 consecutive  seek calls without reading.
+    in.seek(BUF_SIZE * 2, SEEK_SET);
+    in.seek(BUF_SIZE, SEEK_SET);
+
+    readIntBinary(x, in);
+    ASSERT_EQ(x, 8);
```
Comment (CurtizJ) [inline src/IO/tests/gtest_seek_backwards.cpp:35]: In current master `80` will be read.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
In other words - when we read zero bytes after seek, we don't change the buffer contents, but it becomes inconsistent with the position in file?

Maybe we can reset buffer to zero when we read zero bytes at the different offset?

**Comment by Felixoid** [comment]
@mergifyo rebase

**Comment by Felixoid** [comment]
@Mergifyio rebase

**Review by alesapin — APPROVED** [review]

**Comment by alesapin** [comment]
image
Sorry but I cannot resist to merge it.

**Comment by CurtizJ** [comment]
@alexey-milovidov 

> Maybe we can reset buffer to zero when we read zero bytes at the different offset?

It's hard because there are two different buffers: `CompressedReadBuffer` and `ReadBufferFromFileDescriptor` and when we request to read zero bytes from `CompressedReadBuffer` we can't easily and safely reset buffer of `CompressedReadBuffer`, which is used by `CompressedReadBuffer`. Anyway, I think that is should be possible to do several consecutive seeks, because such situation may appear in other circumstantiates, not only while reading zero bytes.

**Comment by CurtizJ** [comment]
> Need to check other buffers that allow seeks?

Yes, it's worth to check.

**Pull request merged**